subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/ProviderCodecs.kt

```diff
@@ -116,7 +124,15 @@ class FixedValueReplacingProviderCodec(
             }
             1.toByte() -> ValueSupplier.ExecutionTimeValue.missing<Any>()
             2.toByte() -> ValueSupplier.ExecutionTimeValue.ofNullable(read()) // nullable because serialization may replace value with null, eg when using provider of Task
-            3.toByte() -> ValueSupplier.ExecutionTimeValue.changingValue<Any>(providerWithChangingValueCodec.run { decode() }!!.uncheckedCast())
+            3.toByte() -> {
+                val value = read()
+                @Suppress("UNCHECKED_CAST")
+                val sideEffect = read() as ValueSupplier.SideEffect<in Any>
```

Comment:
💄 `val sideEffect = readNonNull<ValueSupplier.SideEffect<in Any>>()`

Reply:
Done